eyeball-im-util/src/vector/limit.rs

```diff
@@ -0,0 +1,367 @@
+        }
+
+        // Now, let's poll `VectorDiff`s from the `inner_stream`.
+        let Some(diffs) = ready!(self.inner_stream.as_mut().poll_next(cx)) else {
+            return Poll::Ready(None);
+        };
+
+        // Now, let's consume and apply the diffs if possible.
+        diffs.for_each(|diff| self.apply_diff(diff));
+
+        // If any value is ready, let's return it.
+        match self.ready_values.pop_front() {
+            Some(diff) => Poll::Ready(Some(diff)),
+            None => Poll::Pending,
+        }
+    }
+
+    fn apply_diff(&mut self, diff: VectorDiff<VectorDiffContainerStreamElement<S>>) {
+        let limit = *self.limit;
+        let length = self.buffered_vector.len();
+
+        // Let's update the `buffered_vector`. It's a replica of the original observed
+        // `Vector`. We need to maintain it in order to be able to produce valid
+        // `VectorDiff` when items are missing.
+        self.update_buffered_vector(&diff);
+
+        // If the limit, we have nothing to do.
+        if limit == 0 {
+            return;
+        }
+
+        let is_full = length >= limit;
+        let has_values_after_limit = length > limit;
+
+        match diff {
+            VectorDiff::Append { mut values } => {
+                if is_full {
+                    // Let's ignore the diff.
+                } else {
+                    // Let's truncate the `values` to fit inside the free space.
+                    values.truncate(min(limit - length, values.len()));
+                    self.push_ready_value(VectorDiff::Append { values });
+                }
+            }
+            VectorDiff::Clear => {
+                self.push_ready_value(VectorDiff::Clear);
+            }
+            VectorDiff::PushFront { value } => {
+                if is_full {
+                    // Create 1 free space.
+                    self.push_ready_value(VectorDiff::PopBack);
+                }
+
+                // There is space for this new item.
+                self.push_ready_value(VectorDiff::PushFront { value });
+            }
+            VectorDiff::PushBack { value } => {
+                if is_full {
+                    // Let's ignore the diff.
+                } else {
+                    // There is space for this new item.
+                    self.push_ready_value(VectorDiff::PushBack { value });
+                }
+            }
+            VectorDiff::PopFront => {
+                self.push_ready_value(VectorDiff::PopFront);
+
+                if has_values_after_limit {
+                    // Push back a new item.
+                    self.push_ready_value(VectorDiff::PushBack {
+                        // SAFETY: It's safe to `unwrap` here as we are sure a value exists at index
+                        // `limit - 1`. We are also sure that `limit > 1`.
+                        value: self.buffered_vector.get(limit - 1).unwrap().clone(),
+                    });
+                }
```

Comment:
Even if we are sure that a value at index `limit` exists, I would still prefer to panic with `unwrap` in case of, instead of having a silent error.

Is the idea to remove a `SAFETY`/`unwrap` block?

Reply:
The idea is that it's simpler and doesn't require `unwrap`. The `get()` call will return `Some()` if `has_values_after_limit` is `true`, and `None` if it's `false`. So we don't really need that here. It's only the `PopBack` below that needs it, maybe we should actually inline the definition there?